taxi_demo/scoring/run-model.py

```diff
@@ -0,0 +1,131 @@
+import boto3
+import cloudpickle
+import io
+import logging
+import os
+import pandas as pd
+
+from botocore.client import ClientError
+from flask import Flask
+from flask import jsonify
+from flask import request
+from marshmallow import fields, Schema
+
+app = Flask(__name__)
+
+logger = logging.getLogger()
+logging.basicConfig(
+    format="%(levelname)s [%(asctime)s] %(message)s",
+    datefmt="%Y-%m-%d %H:%M:%S"
+)
+
+
+def _get_env_var(var_name: str) -> str:
+    try:
+        return os.environ[var_name]
+    except KeyError:
+        msg = f"Environment variable '{var_name}' must be set"
+        logger.fatal(msg)
+        raise RuntimeError(msg)
+
+
+MODEL_BUCKET = _get_env_var("MODEL_BUCKET")
+MODEL_FILE = _get_env_var("MODEL_FILE")
+
+logger.info("Fetching model file")
+s3 = boto3.client("s3")
+
+stream = io.BytesIO()
+
+try:
+    logger.info("Downloading model object")
+    s3.download_fileobj(
+        Bucket=MODEL_BUCKET,
+        Key=MODEL_FILE,
+        Fileobj=stream
+    )
+    logger.info("Successfully downloaded model object")
+except ClientError as err:
+    msg = "Failed to download model: {}".format(err)
+    logger.fatal(msg)
+    raise RuntimeError(msg)
+
+logger.info("Deserializing model")
+stream.seek(0)
+model = cloudpickle.loads(stream.read())
+del stream
+logger.info("Successfully deserialized model")
+
+
+class InputSchema(Schema):
+    """
+    Expected schema for input records. In the future, this
+    might be stored along with the model and retrieved at runtime.
+
+    For details on timestamp fields in marshmallow, see
+    https://marshmallow.readthedocs.io/en/stable/_modules/marshmallow/fields.html
+    """
+    passenger_count = fields.Int(required=True)
+    tpep_pickup_datetime = fields.DateTime(required=True, format="iso8601")
+    pickup_taxizone_id = fields.String(required=True)
+    dropoff_taxizone_id = fields.String(require=True)
+
+
+@app.route('/api/model-info', methods=['GET'])
+def get_info():
+    """
+    Returns the following details that applications using
+    the model might need:
+
+    * the expected schema for inputs
+    """
+    out = {
+        "fields": {
+            field_name: type(value).__name__
+            for field_name, value
+            in InputSchema.__dict__["_declared_fields"].items()
+        }
+    }
+    return(jsonify(out))
+
+
+@app.route('/api/predict', methods=['POST'])
+def predict():
+    """
+    This method expects a JSON payload with the structure
+    defined in ``InputSchema``.
+
+    Returns:
+
+    .. code-block:: JSON
+
+        {
+            "prediction": <number>
+        }
+    """
+    global model
```

Comment:
Oh interesting. So this loads at the top of the file and then it is ready to be read anytime? `global` makes me kind of nervous but I guess this is safe.

Reply:
yeah exactly. The other alternative is re-reading the model every time you make a prediction, which will be real slow.

This is just a thing you have to do because Flask's interface is functional. If we used something else where you implement handler classes, the model could be loaded once and stored at `self.model`.

I just know Flask the back and it was the quickest thing to write, yenno.

Reply:
I also think this kind of pattern is a lot less worrying in app code (which only runs here in this lil sandbox) than it would be in library code that's called in an unpredictable environment

Reply:
yep yep